### ISSUE TITLE
I don't quite get how this works

### ISSUE THREAD

**Issue by armin-reichert** [webfx-project/webfx-cli#8]
Say I want to use webfx to build a GWT executable (and maybe others too) for an existing JavaFX application (in my case: https://github.com/armin-reichert/pacman-javafx)

Do I have to bring my Maven artifacts to some central Maven repository first to get this working, or can webfx also resolve dependencies against my local Maven repository?

**Comment by salmonb** [comment]
Nice game! But WebFX doesn't support 3D. So the first step would be to make a 2D version only of the game (otherwise GWT will complain). If you can prepare this (in a separate branch or repository), I can help you making a WebFX version of it.

FYI the Maven libraries must be listed in the webfx.xml file, but the syntax is not yet documented...

Let me know if you're happy to make that 2D version.

**Comment by armin-reichert** [comment]
The 2D and 3D views of my game are already rather well separated, so it will not take a big effort for me to create a 2D-only version. I have already started into this direction. As soon as I have finished this task I will let you know.

I really appreciate your help in creating the webfx version of my Pac-Man/Ms. Pac-Man games. Hopefully it will make your (still small but nevertheless impressive) list of demo apps even more attractive.

Regards
Armin

**Comment by armin-reichert** [comment]
Salu Bruno,

in the meantime I have split the project into a 2D-only and a 2D+3D variant:

- https://github.com/armin-reichert/pacman-javafx/tree/main/pacman-ui-fx (2D only)
- https://github.com/armin-reichert/pacman-javafx/tree/main/pacman-ui-fx-3d (2D+3D, dashboard etc.)

The script `build.bat` in the repository root directory builds everything. The `run.bat` starts the full version and the `run2d.bat` script runs the 2D-only version. Alternatively, the "shaded" jar files in the target directories can be used.

Now I would be very happy if you could help in creating

1. A browser version of the 2D-only game (Prio 1)
2. Executables/installers for Windows/Linux/Mac for the full game (Prio 2)

What do you think?

Best regards
Armin

**Comment by salmonb** [comment]
Thanks, I will have a look at it this evening after my workday.

If we achieve 1), we achieve 2) at the same time, as WebFX can create these installers by default. WebFX is also compatible with Gluon, so we will get the mobile versions at the same time (but only the Android version is easy to share as Apple makes things complicated to share ad-hoc apps).

For example here are the resulting artifacts of the [JArkanoid](https://github.com/webfx-demos/webfx-demo-jarkanoid/releases) game (just expand the assets to see them). If we manage to port it to WebFX, we will have the same for your game.

**Comment by armin-reichert** [comment]
Sounds great. So "enjoy" your work day ;-)

**Comment by salmonb** [comment]
I made the WebFX port, so now your game is playable online! https://pacman.webfx.dev 😃 

I made these 2 forks in the WebFX demos space:
- https://github.com/webfx-demos/webfx-demo-pacman-core (fork of your pacman-basic)
- https://github.com/webfx-demos/webfx-demo-pacman (fork of your pacman-javafx)

I had to make quite of lot of changes (mainly a backport to java language level 11 for GWT + disabling some features such as the score file persistence - I can write a score persistence later that will work in the browser).

The main issue is that the game is on the left side of the screen. In the browser, you can't control the size of the window, so you need to update your code so the game looks centred even if the window is wider. Can you fix that in JavaFX and let me know when it's ready?

**Comment by armin-reichert** [comment]
Hello Bruno,

thank you very much for your effort to get this first version runnning!

I understand that there is still much room for improvement. In which branch/repository should I make the needed changes?

About the layout issue: I don't understand why the layout is left-aligned. In a window of the desktop version the canvas is centered if the window is wider. Do you have any idea what I should change to make the browser happy?

Please do not publish the game on your demo site until we have fixed the most annoying issues.

Best regards
Armin

**Comment by armin-reichert** [comment]
I think just changing the background will not help: if you have a close look at the intro scene, the ghosts come from the right into the visible area. But as they would normally not be visible when outside of the subscene and now they are, this looks strange (hope you understand what I mean). Of course I could also change the intro scene code such that the ghosts become visible only when they enter the subscene area.

**Comment by salmonb** [comment]
Yes, I fixed the issue by giving the canvas a constant size (WIDTH, HEIGHT) and replacing 1) canvas.getTransform().add(scale) with just 2) canvas.setScaleX/Y(scaling). You don't need to resize the canvas as you just scale it. And 1) and 2) are not doing the same regarding the pivot. 2) is always taking the centre as the pivot, which is what we want.

For the sound are you talking about the application start? Because in the browser sandbox, it's not possible to play sound before the user interacted.

**Comment by armin-reichert** [comment]
Hi,

no I mean the cutscene sound, They are played after completing level 2, 5, 9, 13 etc. They have no sound yet.

There is also a key combination Alt+Z that when pressed on the intro screen should play all cutscenes. Any idea why this shortcut is not working?

**Comment by salmonb** [comment]
Try Alt-W instead (a keyboard mapping issue, will try to fix that in WebFX).

Can you tell me where is that code that plays the cutscenes sound?

Yes, StackPane is supported, but I will have a look at the flash messages.

**Comment by salmonb** [comment]
I fixed the sound issue: it was caused by some missing breaks in switch cases when I did the backport to java 11 -> GameModel.SE_STOP_ALL_SOUNDS case was executed just after the cutscene sound...

For the keyboard, this page displays all the info we can get from JavaScript: https://www.toptal.com/developers/keycode (just type the key combination you want)

**Comment by armin-reichert** [comment]
Merveilleux :-) 

However, the sound in PacMan cutscenes 1 and 3 is still not ok, the sound in that scenes should be repeated twice. Flash messages work now too (ok, I would like a sans-serif font better :-). 

Alt+PLUS (Code 187) and Alt-MINUS (Code 189) however do not work, they are used to increase/decrease the game loop speed and this feature is important (I saw a question on Reddit where someone asked exactly for this feature for a handicapped friend of him).

I will test further and let you know of my findings.

Thanks again very much for your work!

Regards
Armin

**Comment by armin-reichert** [comment]
Formatting of locale dependent texts from text resource bundle does not work. Press Alt+L (cheat: add 3 lives) during game play, then the flash message text is not formatted.

**Comment by salmonb** [comment]
The issue I was talking about was not with Alt, but with what code will represent the + or - keys. I did a test by changing `alt(KeyCode.PLUS)` to `alt(KeyCode.ADD)`  & `alt(KeyCode.MINUS)` to `alt(KeyCode.SUBSTRACT)` and it's working on my computer, but if I commit that, it will stop working for people who don't have a numpad... The KeyCode is quite specific to the keyboard layout, that's why I was suggesting you could keyEvent.getText() instead, or eventually allow several key combination (PLUS or ADD, MINUS or SUBSTRACT, etc...) to solve this issue.

**Comment by salmonb** [comment]
Yes, I disabled text formatting, as it's not supported by GWT. I will see what I can do.

**Comment by armin-reichert** [comment]
I see. As mentioned, we can just change the Alt+PLUS, Alt-MINUS, Alt-DIGIT0 to anything less troublesome.

Text formatting is completely unimportant, I can just format the text with String.formatted()  instead of using the resource bundle patterns like `{0,number,integer}`.

**Comment by salmonb** [comment]
String.formatted() is not supported by GWT neither. But don't worry, I will find a quick fix for this case

**Comment by armin-reichert** [comment]
Ok, we are in Java 11. Well there are StringBuilder, StringBuffer or good old string concatenation :-)

**Comment by salmonb** [comment]
Format issue fixed. I will have a look at the warning.

Yes, your idea is good to offer an alternative to the sound as we can't play it from the beginning in the browser.
Maybe it can be just a background image, a bit like they did for SpaceFX https://spacefx.webfx.dev

**Comment by salmonb** [comment]
Warnings fixed.

**Comment by armin-reichert** [comment]
Yes, I need a start screen explaining the most important key combinations and offering a menu for selecting which game to play.  

Maybe in the browser version one should also add some black padding around the game area, like here: https://www.novumgames.com/spiel/ms-pacman/ 
(seems to be based on Shaun Williams Javascript implementation https://github.com/masonicGIT/pacman)

There should also be a key combination to activate the cheats.

Aside from all these little things, I have to say that I am really impressed by your project!

Regards
Armin

**Comment by salmonb** [comment]
Thanks. I was myself impressed by the good results I could get when I was developing WebFX. GWT is amazing. And OpenJFX as well. A good marriage!

I just applied an additional 0.95 factor to the canvas scale to create the padding. Does that look ok?

**Comment by armin-reichert** [comment]
Good morning Bruno,

yes, the margin around the canvas looks fine to me, however a black padding around the canvas on each side would even look better. But this has no high priority. 

What is more important to me is the "signature" saying "A Fan Tribute...". I intentionally placed this "handwritten" signature in both games right above the original copyright notice to clearly express that this clone is not meant to be a copyright violation. 

If there is any technical restriction hampering the signature being shown like this, I could replace the text by an image and place that image exactly there (I use an overlay pane over tha canvas to place the signature to avoid the text getting blurred).

Another general question: If I add additional fixes or features like the mentioned start screen, how will these changes be merged into your fork/branch?

**Comment by salmonb** [comment]
Guten Morgen Armin,

The previous games I ported to WebFX were finished, so I didn't have this problem. But we can try and see how easy or difficult it will be to resync the WebFX version from your version.

So just continue the development in your repository as normal (you can continue using recent Java features) and let me know when you reach a stage you would like to move to the WebFX version. Then I will resync my fork and get your modifications in the main branch of the fork, the WebFX version is in a separate webfx branch. Then I will see if a Git rebase can do miracles or not...

**Comment by armin-reichert** [comment]
Salu Bruno,

I don't want to bother you too much with this (and I think you have more than enough real work to do). Would it be possible that I create a fork(?) of your repositories and integrate my future changes in my local copy, run the toolchain and check if everything works? And when I have finished everything I want to have in the showcase version, we could reintegrate the changes into your repository again?

**Comment by salmonb** [comment]
That's another possible option indeed, and it will work for sure. It's just more work for you (and less for me! 😉). You can fork my fork(s), make the changes, and submit them to me as a pull request. Let me know if you have difficulties to build the webfx version on your local machine.

**Comment by salmonb** [comment]
Lucky one!

I'm sure this doesn't happen, but let me know if you get bored, I can give you some jobs! 😄

Ex: a backport tool that can transform java code back to 11 (unless that already exists and I don't know...)

**Comment by armin-reichert** [comment]
Concerning the backport tool: IIRC, In the 90s, phone books on CD were sold in the millions. In Germany however, it was illegal to sell CDs with a backwards search, i.e. provide the phone number and find the owner of that number (it is legal now). Some clever company outsourced the job to China, let some poor(?) Chinese people type all the German phone books into a database and create CDs with inverse search from that. At least that is my recall of the story. That should give you an idea for such a "tool" ;-)

**Comment by salmonb** [comment]
If you start from scratch, that would be a huge task indeed, but there are some generic tools available now to analyse and transform Java code, such as [Spoon](https://spoon.gforge.inria.fr) (someone mentioned it in this [issue](https://github.com/webfx-project/webfx-cli/issues/1)).

**Comment by salmonb** [comment]
Yes, Google now focuses on [J2CL](https://github.com/google/j2cl), the future replacer of GWT. As mentioned in the [roadmap](https://github.com/webfx-project/webfx/blob/main/.github/ROADMAP.md), I will move to J2CL when it will be officially out. But things take time, so in the meantime, I have to stay with GWT, which is stuck to Java 11.

**Comment by armin-reichert** [comment]
Good morning Bruno,

when I try to fork your fork https://github.com/webfx-demos/webfx-demo-pacman, GitHub complains and claims that my original repository already is a fork of your fork and no more forks are allowed. 

This does not match my mental model of a fork, but I am far from understanding git. So do you have any idea what the problem is?

Maybe we should "cut the connection" of your forked repositories to the original ones? Then I probably could fork them and create pull requests for the handful of changes I want to make before publishing the game on your website.

Does that make any sense?
Armin

**Comment by salmonb** [comment]
Weird, but no worries, I gave you a direct write access to the repo. The port is located in the webfx branch. Be aware that each time you will push your work into that branch, this will automatically update https://pacman.webfx.dev (after about 5 min, thanks to a GitHub workflow)

**Comment by armin-reichert** [comment]
Thanks! So I cloned the `webfx-demo-pacman` repository to my local machine and checked-out the webfx branch. Local build with the webfx cli works. 

Now I have one question: To test the GWT version locally, is it sufficient to install a local appserver (like Tomcat) and "deploy" the war file, or is there some other preferred way in 2023? (The last time I worked with Tomcat was around 20 years ago )

**Comment by salmonb** [comment]
You can run the demo with `webfx run -g` but this will just open the html file in the browser (you can locate the html file with `webfx run -gl`), but sound is not allowed by the browser in this case. I usually open the html file in IntelliJ IDEA (my Java IDE), as it serves it through its internal web server, and then sound is working in this case.

**Comment by armin-reichert** [comment]
Sorry, just as I wrote the comment I saw it in the webfx documentation.

I have now everything locally in my IDE (Eclipse). I see an error in the -gwt subproject (which I can probably ignore), everything else is ok.

Some first findings: Eclipse shows warnings because of missing `requires transitive` in the generated module-info.java. Not sure if this is an issue but maybe your code generator should generate them wherever they make sense?

**Comment by salmonb** [comment]
Great.

Not sure about the Eclipse warning, the essential is that it compiles ok with the Maven build chain.

During the development phase, you can run and debug the OpenJFX version in your IDE. The module to run is `pacman-ui-fx-openjfx` and the main class is `dev.webfx.platform.boot.ApplicationBooter` (as always with WebFX)

**Comment by armin-reichert** [comment]
One question: I pushed now 2 commits to the webfx branch of `webfx-demos/webfx-demo-pacman`. Do I additionally have to create a pull request to start the central build and update the demo website?

**Comment by salmonb** [comment]
Nothing to do, it's all automatic 😄 

You can see if the GitHub workflow ran ok here: https://github.com/webfx-demos/webfx-demo-pacman/actions

This workflow regenerates the demo website + all artifacts in the release https://github.com/webfx-demos/webfx-demo-pacman/releases

**Comment by armin-reichert** [comment]
Hm, I see a message "webfx had recent pushes 14 minutes ago" on the GitHub page of "webfx-demos/webfx-demo-pacman". And a button "Compare & pull request" so I would think there is manual interaction needed?

**Comment by salmonb** [comment]
Just ignore this message, as our case is a bit specific: we don't want the webfx branch to be merged back to the main branch.

**Comment by armin-reichert** [comment]
Thanks. On `https://github.com/webfx-demos/webfx-demo-pacman/actions` I see now that my two commits have been processed. So I will not bother you further as you are the one who has real work to do ;-)

**Comment by armin-reichert** [comment]
Sorry to bother you again. Things don't work as expected :-(

I have all webfx tools available on my machine. Creating a "hello world" sample, building/runnning it with `webfx build -fr`, `webfx build -gr` and `webfx build -dr` all work as expected and the hello world app opens, either in a window or in the browser.

But for the `webfx-demos/webfx-demo-pacman`, the app only opens in the browser! For the other variants, no window opens. 

Could you please verify that this is the same in your local version? It cannot be caused by my 3 changes so far because I only changed the window background to black, added gitignore files and removed the 3D subproject.

**Comment by salmonb** [comment]
Here are the results on my iMac:
- `webfx build -gr` 👍 
- `webfx build -fr` 👍 
- `webfx build -dr` 👎  (compiles but crashes - it's actually _normal_ because Gluon doesn't support JavaFX media yet on Windows and MacOS as stated in their [doc](https://docs.gluonhq.com/#:~:text=Table%203.%20Platform%20feature%20matrix%20for%20Native%20Images))

BTW if you want a mobile version of your game, it's possible but I will need to modify the code a little bit so playing sound can be processed by Gluon Attach on Android & iOS (the desktop & web version will still work after that modification). Let me know if you're interested.

I'm surprised that `webfx build -fr` doesn't work for you. Do you have any error message in the console?
Have you tried to run the Windows fat jar and installer exe generated by the GitHub workflow (available [here](https://github.com/webfx-demos/webfx-demo-pacman/releases))?

**Comment by salmonb** [comment]
hi Armin,

How is it going?

I just noticed that the latest online version is not working, but this is probably due to the following WebFX bug that I would like to make you aware of: when changing a node using parent.getChildren().set(x, newNode), the WebFX mapper may get it wrong and put the node at the end instead of position x.

I will try to fix that bug later, but for the time being, is it possible for you to reset the whole list instead?
Ex: parent.getChildren().setAll(node1 ,node2, etc...)?

**Comment by armin-reichert** [comment]
Of course I will tell you about any strange effects I find.

Example: I have the impression that mouse clicks on a Text widget do not work properly. I had a Text instance only for the greeting message ("Click to start") but that did not work in the browser, so I had to put the text inside a panel.

I lost some time because I had to install IntelliJ to work with the webfx branch (I had absolutely no idea how to realize the "application configuration" involving ApplicationBooter inside Eclipse). But working with IntelliJ is really fun, Git integration is top.

**Comment by salmonb** [comment]
Yes, IntelliJ is amazing, I 💟 it too

The text only (without pane) should work... I just pull the latest source, and noticed a call to `greetingText.setMouseTransparent(true);` which prevents any interaction with the user. Maybe that's the problem?

I'm glad that you have the OpenJFX configuration running in IntelliJ, so you can now see if your issues come from WebFX or the application code. For example, if the text issue is because of setMouseTransparent(), you should see that the issue is also happening with OpenJFX, and therefore it's not a WebFX bug. This OpenJFX configuration should really accelerate your development cycle! 😃

**Comment by armin-reichert** [comment]
Well, I added the `setMouseTransparent(true)` call *after* putting the text inside a StackPane. First I had the text only, but the mouse click on the text was only handled in the desktop version, but not in the browser version. Then I added the surrounding StackPane and wanted the mouse click over the text also being handled on the stack pane.

**Comment by salmonb** [comment]
I tried this locally, and it works in the browser:

```java
	private Node createGreetingPane() {
		DropShadow ds = new DropShadow();
		ds.setOffsetY(3.0f);
		ds.setColor(Color.color(0.2f, 0.2f, 0.2f));

		var greetingText = new Text(">Click to start<");
		//greetingText.setMouseTransparent(true);
		greetingText.setEffect(ds);
		greetingText.setCache(true);
		greetingText.setFill(Color.YELLOW);
		greetingText.setFont(AppRes.Fonts.font(AppRes.Fonts.arcade, 24));
		//var pane = new StackPane(greetingText);
		greetingText.setOnMouseClicked(e -> {
			layers.remove(LAYER_GREETING);
			rebuildMainSceneLayers();
			simulation.start();
			Actions.playHelpVoiceMessageAfterSeconds(4);
		});

		//return pane;
		return greetingText;
	}
```

**Comment by armin-reichert** [comment]
Definitely didn't work here (Windows10, Google Chrome). I could try it again with the exact code above.

**Comment by salmonb** [comment]
Maybe it's a Windows thing (as I don't test on Windows...) If you can test for me, that would be helpful.

PS: Good idea the initial click, as it makes a user interaction, and then the sound works in the browser 👍 

Tips: if you don't want that greeting text to show in the other platforms (because they can play sound straightaway), you can use the condition `dev.webfx.platform.useragent.UserAgent.isBrowser()` (just run `webfx update` after inserting this in your code and the WebFX CLI should automatically insert the required WebFX module - and after that IntelliJ should propose you to replace the fully qualified name with a simple `UserAgent.isBrowser()` + `import dev.webfx.platform.useragent.UserAgent;`).

**Comment by armin-reichert** [comment]
Yes, that "no sound without initial gesture" feature was the reason I added this "Click Me" thing. Thanks for the tip, but the webfx branch should just be used inside the browser. 

You mentioned that webfx can also be used to build executables for different platforms. That's exactly what I am looking for. But will this also only be possible after downporting to Java 11 and all these changes you had made?

**Comment by salmonb** [comment]
WebFX can compile apps to many platforms (not only the web) all from a single source code (your code in the webfx branch). The current GitHub workflow already generates different installers of your game as you can see in the generated assets [here](https://github.com/webfx-demos/webfx-demo-pacman/releases) (just expand the Assets).

If you have an Android device, you can try to download and install the .apk file, but it's probably crashing because of JavaFX media not supported by Gluon. However, I can easily solve this problem using the WebFX API (let me know if you're interested).

Once everything works with WebFX, we can try to back port that into your original game.

**Comment by armin-reichert** [comment]
Hi Bruno,

that was not a call for action, I just collect what I find on my way and let you know.

**Comment by salmonb** [comment]
Why are you not testing the web version on your local machine before pushing? This should be must faster.

I will have a look at the text positioning this evening.

**Comment by armin-reichert** [comment]
Sorry I got a little bit confused (happens at age 56). When I installed IntelliJ some days ago I quickly clicked through the installation and overlooked that IntelliJ uses its own local Git root. My console window still referred to the Git root I use with Eclipse so the `webfx build -gr` obvioulsy had no effect ony the changes I made with IntelliJ. Stupid me.

But now, when I refer to the IntelliJ Git root and call `webfx build -gr` and the browser opens (via a file: URL) I observe that some things like tihe boot "animation" do not work properly. Maybe I must let the page being served by a local webserver or use the one inside IntelliJ as you proposed.

**Comment by salmonb** [comment]
I hadn't spotted the spritesheet issue, I see now what you mean, but it happens only when opening from `file://`, so not an issue for the final web version. Have you found how to open the html file with the IntelliJ built-in webserver? 

Another 💡: You can bookmark the generated index.html in IntelliJ so you can access it quicker next times.

**Comment by salmonb** [comment]
Ok, the positioning issue is with the JavaFX `textOrigin` property. For now WebFX supports VPos.TOP, VPos.CENTER and VPos.BOTTOM, but not VPos.BASELINE, which is unfortunately the default value. I will investigate again if there is a way to support VPos.BASELINE in HTML, but in the meantime, you can just set a different value to textOrigin and adjust your y value accordingly. Let me know if that solves the issue.

**Comment by salmonb** [comment]
Actually I think I found a solution and fixed the VPos.BASELINE html mapping in WebFX!

So no need to change your code finally, just get the latest webfx version (your Maven will automatically load it tomorrow, but if you don't want to wait, just clean this folder in your Maven local repository (located here from your user home: `.m2/repository/dev/webfx/webfx-kit-javafxgraphics-peers-gwt`)

**Comment by armin-reichert** [comment]
Works!

**Comment by salmonb** [comment]
Great, and I triggered the GitHub workflow manually again (as no change is needed in your source code), so now the online version works as well

**Comment by armin-reichert** [comment]
Hi Bruno,

I have an issue with a GridPane. For the context-sensitive help/menus (press 'H' to open/close) I use a GridPane together with a surrounding HBox to give a transparent background and some padding. In the desktop version that looks perfect, but in the browser, the vertical aligment is broken. Any idea?

**Comment by salmonb** [comment]
Looks like a similar issue as yesterday, probably my fix for VPos.BASELINE doesn't work in all scenarios... I will have a look this evening, but for the time being, just try another value for the textOrigin. For example try `text.setTextOrigin(VPos.TOP);`  and you should have a better result.

**Comment by salmonb** [comment]
hi Armin, how is it going?

FYI I just pushed some fixes in WebFX, this includes the text positioning (so you might need to remove the translation correction now).

Also as you probably noticed, I fixed the background positioning issue. But on my big 27" screen, the picture is so stretched in width that I see only the top of the image (and the yellow "Click to start !" is on top of the yellow image). I think it would be better to put that image in 'contain' mode (instead of 'cover'), and put the wallpaper behind, so something like:

```java
var bgImage = new BackgroundImage(
			AppRes.Graphics.msPacManCabinet,
			BackgroundRepeat.NO_REPEAT,
			BackgroundRepeat.NO_REPEAT,
			BackgroundPosition.CENTER,
			new BackgroundSize(AUTO, AUTO, false, false, true, false));
		root.setBackground(new Background(
				new BackgroundImage(AppRes.Graphics.wallpaper, null, null, null, null),
				bgImage));
```

What do you think?

**Comment by armin-reichert** [comment]
Hi Bruno,

I am making progress. Working mode with IntelliJ and „webfx cli“ is fluent, building the GWT version locally takes around 20 seconds, that’s ok for me.

The greeting page is still provisionary, I probably will select another wallpaper, something like https://wallpapercave.com/w/wp4163225

But I am not sure about copyright etc. The assets in my games are also „stolen“ from other projects. But how should one clone a classic Arcade game without using the original assets?

Yes, I noticed that you already fixed some of the alignment issues. That’s great!

What I want to add in any case is some black area around the game scene, something like here:
https://www.pacman1.net/

But I am open to your ideas of course too. There is not much missing until we can publish it.

Best regards
Armin




Von: Bruno Salmon
Gesendet: Samstag, 13. Mai 2023 14:35
An: webfx-project/webfx-cli
Cc: Armin Reichert; Author
Betreff: Re: [webfx-project/webfx-cli] I don't quite get how this works (Issue#8)

hi Armin, how is it going?
FYI I just pushed some fixes in WebFX, this includes the text positioning (so you might need to remove the translation correction now).
Also as you probably noticed, I fixed the background positioning issue. But on my big 27" screen, the picture is so stretched in width that I see only the top of the image (and the yellow "Click to start !" is on top of the yellow image). I think it would be better to put that image in 'contain' mode (instead of 'cover'), and put the wallpaper behind, so something like:
var bgImage = new BackgroundImage(
			AppRes.Graphics.msPacManCabinet,
			BackgroundRepeat.NO_REPEAT,
			BackgroundRepeat.NO_REPEAT,
			BackgroundPosition.CENTER,
			new BackgroundSize(AUTO, AUTO, false, false, true, false));
		root.setBackground(new Background(
				new BackgroundImage(AppRes.Graphics.wallpaper, null, null, null, null),
				bgImage));
What do you think?

**Comment by salmonb** [comment]
For your black area, you can embed the canvas in a black StackPane container and give it a size slightly superior (ex: +10%), so something like this:

```java
	protected GameScene2D(GameController gameController) {
		checkNotNull(gameController);
		context = new GameSceneContext(gameController);

		StackPane canvasPane = new StackPane(canvas);
		canvasPane.setBackground(new Background(new BackgroundFill(Color.BLACK, new CornerRadii(10), null)));
		double marginFactor = 1.1;
		canvasPane.setMaxWidth(WIDTH * marginFactor);
		canvasPane.setMaxHeight(HEIGHT * marginFactor);

		root = new BorderPane();
		root.heightProperty().addListener((py, ov, nv) -> {
			double scaling = nv.doubleValue() / (HEIGHT * marginFactor);
			canvasPane.setScaleX(scaling);
			canvasPane.setScaleY(scaling);
			// don't ask me why this works but setScaleX/Y doesn't
			overlay.getTransforms().setAll(new Scale(scaling,scaling));
		});

		canvas.setWidth(WIDTH);
		canvas.setHeight(HEIGHT);

		// help appears in overlay layer at left scene border, 10% from top
		helpRoot.setTranslateX(8);
		helpRoot.setTranslateY(HEIGHT * 0.1);

		friendlyGhostImage = new ImageView(AppRes.Graphics.friendlyGhostIcon);
		friendlyGhostImage.setPreserveRatio(true);
		friendlyGhostImage.setFitHeight(20);
		friendlyGhostImage.setTranslateX(-20);
		friendlyGhostImage.setTranslateY(helpRoot.getTranslateY());
		friendlyGhostImage.setOnMousePressed (e -> Actions.showHelp());

		overlay.getChildren().addAll(helpRoot, friendlyGhostImage);

		layers.getChildren().addAll(canvasPane, overlay);
		root.setCenter(layers);

		infoVisiblePy.bind(Env.showDebugInfoPy);
	}
```

You will need to remove the round corners from the canvas as they will be now in that container.

**Comment by salmonb** [comment]
hi Armin, just to let you know, I created a new `webfx-bundle` branch that you can merge into the webfx branch if you want (and delete `webfx-bundle` once done). It does load the bundles from `messages.properties` (sorry for the dirty hardcoding in the initial version). FYI it's necessary to list messages.properties as an embed resource in webfx.xml to make this work (otherwise loading the resource would be asynchronous) and run `webfx update`.

Just a suggestion for the font used in the help: wouldn't it be better to use a monospace font (ex: Courier) to get the shortcuts aligned in the right part?

**Comment by armin-reichert** [comment]
Hi Bruno,

thanks for the effort. Message bundles are not so important for this demo but I will do what you suggested. 

What about the persistent high scores? Would it be much work to make these work too?

**Comment by salmonb** [comment]
`GridPane.setHgap()` and `GridPane.setVgap()` do have effect (try with a big value like 100 and you will see). Maybe it's more a difference between the font metrics between JavaFX & HTML? I will have a deeper look later...

For the high scores persistence, yes it's possible. I will create a new branch for this that you will merge to webfx. Will let you know once done.

**Comment by armin-reichert** [comment]
I just pushed some changes (monospace font, workaround for grid gap issue). 

BTW: Yesterday, I observed differences between my local GWT version and the version created by the build on the website. Do I have to update something locally to get them in sync again?

**Comment by salmonb** [comment]
That's probably because I pushed my fixes on WebFX yesterday (that's why I sent you a message). This is immediately considered on the GitHub workflow because the workflow rebuilds a new VM from scratch and therefore always download the latest Maven snapshots. But on your local machine, Maven does a snapshot refresh only once a day (on first invocation). So your local WebFX snapshots were not the same in the meantime.

💡 Tip: you can use `webfx update -c` to force the snapshot cleaning (as mentioned in `webfx update -H`)

**Comment by armin-reichert** [comment]
I just merged webfx-bundle into my local webfx branch. Now I get errors:
`Package 'dev.webfx.platform.useragent' is declared in module 'webfx.platform.useragent', but module 'pacman.ui.fx' does not read it`

What to do?

**Comment by salmonb** [comment]
`webfx update` is always a good try

**Comment by armin-reichert** [comment]
At least it transforms compile errors into runtime errors :-)
`Exception in thread "Thread-1" java.lang.RuntimeException: Exception in Application init method
	at javafx.graphics@19/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:896)
	at javafx.graphics@19/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:196)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 4: file:\C:\Users\Armin\IdeaProjects\webfx-demo-pacman\pacman-ui-fx\target\classes\de\amr\games\pacman\ui\fx\assets\texts\messages.properties
	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:232)
	at java.base/jdk.internal.module.Resources.toSafeFilePath(Resources.java:143)
	at java.base/jdk.internal.module.Resources.toFilePath(Resources.java:97)
	at java.base/jdk.internal.module.ModuleReferences$ExplodedModuleReader.find(ModuleReferences.java:382)
	at java.base/jdk.internal.loader.BuiltinClassLoader$2.run(BuiltinClassLoader.java:464)
	at java.base/jdk.internal.loader.BuiltinClassLoader$2.run(BuiltinClassLoader.java:459)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findMiscResource(BuiltinClassLoader.java:458)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findResource(BuiltinClassLoader.java:341)
	at java.base/java.lang.ClassLoader.getResource(ClassLoader.java:1404)
	at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1738)
	at webfx.platform.resource.java@0.1.0-SNAPSHOT/dev.webfx.platform.resource.spi.impl.java.JavaResourceProvider.getResourceInputStream(JavaResourceProvider.java:44)
	at webfx.platform.resource.java@0.1.0-SNAPSHOT/dev.webfx.platform.resource.spi.impl.java.JavaResourceProvider.getText(JavaResourceProvider.java:38)
	at webfx.platform.resource@0.1.0-SNAPSHOT/dev.webfx.platform.resource.Resource.getText(Resource.java:31)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.util.ResourceManager.loadBundle(ResourceManager.java:135)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.app.AppRes$Texts.load(AppRes.java:91)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.app.AppRes.lambda$load$3(AppRes.java:56)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.app.AppRes.load(AppRes.java:62)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.app.AppRes.load(AppRes.java:56)
	at pacman.ui.fx/de.amr.games.pacman.ui.fx.app.PacManGameAppFX.init(PacManGameAppFX.java:73)
	at webfx.kit.openjfx@0.1.0-SNAPSHOT/dev.webfx.kit.launcher.spi.impl.openjfx.JavaFxWebFxKitLauncherProvider$FxKitWrapperApplication.init(JavaFxWebFxKitLauncherProvider.java:102)
	at javafx.graphics@19/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:825)
	... 2 more`

**Comment by salmonb** [comment]
`webfx update` fixed the previous issue. What's happening now is a Windows error when trying to read the file (but your GWT should work). It looks like Windows doesn't like the 'file:' prefix in the path. I will see what I can do...

**Comment by armin-reichert** [comment]
As mentioned, this has no priority at all. Maybe we should just create a list of all the issues found and give them priorities. I really do not want to steal your time with such things.

**Comment by salmonb** [comment]
It's actually good that you report these problems, as I don't have a Windows computer (I have an iMac & a Linux laptop).

I think I fixed the problem (it was in the WebFX Platform layer). Now pushed. Can you run `webfx update -c` and let me know if this solves the issue?

You are my Windows tester 😄

**Comment by armin-reichert** [comment]
Strange. After `webfx update -c` I cannot use the run configuration in IntelleJ anymore. I get `java.lang.ClassNotFoundException: dev.webfx.platform.boot.ApplicationBooter`.

**Comment by salmonb** [comment]
Sometimes IntelliJ gets confused with all these modules. Try "Reload all Maven projects" (the recycling icon in the Maven tool windows on the right). Let me know if it helps.

**Comment by armin-reichert** [comment]
Yes, that helped! Now I can run the (merged) webfx-bundle branch locally without errors, in the VM and in the GWT version. Changes in the `messages.properties` file also appear in the help menu.

However, what do I have to do to get the German bundle `messages_de.properties` at runtime?  In Eclipse, I add the VM parameters `-Duser.language -Duser.country=de` to get that but I didn't success to do that with the WebFX PacMan run configuration in IntelliJ. Or is the local-dependent loading not yet implemented?

**Comment by salmonb** [comment]
Great, this means that the Windows bug is fixed! Thank you! 😃 

You can't pass VM parameters to the browser, but maybe you can use `dev.webfx.platform.windowlocation.WindowLocation.getQueryString()`. Then if you add `?lang=de` to the browser url, that method will return `lang=de` and then you can read `messages_de.properties` instead in `ResourceManager.loadBundle()`. To make this work, you will need to list `messages_de.properties` in the embed resources in `webfx.xml`. Let me know if you need more help with this.

**Comment by armin-reichert** [comment]
I wanted to add the VM parameters to the IntelliJ run config to test it with the VM. How can I do that? The browser is another question. But as I said the resource bundle support has prio lower than low.

**Comment by salmonb** [comment]
Right now the `WindowLocation` methods (such as `getQueryString()`) returns null in the JVM, but maybe I could add a possibility to initialise the window location from VM parameters, and then you could pass `?lang=de` when running in the JVM? In this way the same java code would work for both the browser and the JVM (which is the goal of WebFX). What do you think?

Alternatively, you can add a language menu (EN & DE) in your game, and we persist that setting. But it's more work for you...

Regarding the score persistence, I think I have something ready. Are you ok if I push that in the webfx branch of `webfx-demo-pacman-core`?

**Comment by armin-reichert** [comment]
Hi Bruno,

feel free to make any changes without asking me.

I will continue tomorrow, old men must rest and also have a wine cellar to work on. Malescasse 2009 or Capbern-Gasqueton 2012, that's the question for today :-)

**Comment by salmonb** [comment]
What do you mean old? You said you were 56! I'm not far behind... 🤣 

Ok I pushed the changes, which updated the `webfx-demo-pacman-core` snapshots. You can get them with `webfx update -c` in `webfx-demo-pacman` (this will also update your build chain with a new WebFX module added for the persistence -> you will need include these changes in your next commit & push to GitHub).

A difficult question for the day! But whatever your choice, enjoy! 🍷 And then rest! 🛌 😄

**Comment by armin-reichert** [comment]
1. You can now select the game version via URL query parameter `game`: `game=pacman` or `game=mspacman`. Of course one could also provide different URLs for these entry points.
2. The help button is now inside the frame around the game area and not so distracting anymore.

I am satisfied so far, but maybe you want to test and evaluate too before publishing. What I am not so happy about is the blurred look of the canvas content which is caused by the upscaling. This is more visible in the browser than in the desktop window.

And now I have to rest again ;-)

**Comment by salmonb** [comment]
Wouah, good progress, and great job! 🥳  I will test it tonight 👍 

Regarding your concern with the canvas, I think it's possible to get sharper graphics by replacing the scaling with an actual resize of the canvas (but always keep same width/height ratio). I guess this would require some adaptation of your code to manage an arbitrary canvas size height (you will need to scale your images inside the canvas). I don't know how much work that would represent, depending on how you implemented it...

Do you think it's doable, or is it too much work?

**Comment by armin-reichert** [comment]
I would like to let the scaling (issue) for now as is. (Will have a look at it in my main development branch) From my side, you can publish the stuff now (and spread the good news :-). You might provide two separate URLs / demo page entries, because historically, these are different games. What do you think?

**Comment by salmonb** [comment]
Great news! And nice idea the iPad look! 👍 

I did a bit of testing and there is one little issue with my keyboard: I need to press my A key to quit (instead of my Q key) - I have an AZERTY French keyboard. Some people may have other issues with other keyboard layouts. It's a bit a hassle these keyboards...

I looked at the `KeyEvent` and it looks like the only property that returns the logical key (and not the physical key) is `getText()`(it returns 'a' and not 'q' when I press my A key). Do you think you can consider this property instead of `getCode()` when it's about letters (but not when it's about arrows as getText() returns a String...)? Or do we leave it as it is?

**Comment by salmonb** [comment]
Actually this little trick might solve the problem:

Replacing:
```java
		KeyCode code = e.getCode();
```
with
```java
                 // Trying first to convert the logical key into a key code to consider the keyboard layout
		KeyCode code = KeyCode.getKeyCode(e.getText().toUpperCase());
                 // If not found (probably not a letter), we take the physical key
		if (code == null)
			code = e.getCode();
```

**Comment by armin-reichert** [comment]
Concerning the keyboard, I now understand what you mean: The key event contains key code "KeyCode.Q" if you press "a" on your french keyboard, right? So I will have to the code to handle that situation better. Is this the only of my used key combinations that has that problem?

(I am already working on a fix)

**Comment by salmonb** [comment]
Yes you understood the problem. Yes so far this is only the key combination that doesn't work for me, but like I said, we don't know about other keyboard layouts... But I think my trick (see above as our posts crossed) should fix all issues for all keyboard layouts 🤞

**Comment by armin-reichert** [comment]
I pushed a fix, could you please test?

**Comment by salmonb** [comment]
Yes it works! 🥳 

I was thinking about the 2 game variants: would you prefer to have 1 subdomain for each?
https://pacman.webfx.dev
https://mspacman.webfx.dev

I can easily create the second one and update the GitHub workflow to publish the game in these 2 subdomains at the same time. From your side, you would need to replace your `WindowLocation.getQueryString()` test with `WindowLocation.getHostname()` which will return either `pacman.webfx.dev` or `mspacman.webfx.dev` in the browser (but null for the desktop).

Let me know if you would like this change.

**Comment by armin-reichert** [comment]
Yes. I would prefer two subdomains (with exactly these names) as entry points without the URL parameter. Maybe it's easiest if you make the needed change in the code too (class GameApp:161) where the initial game variant is set.

**Comment by salmonb** [comment]
Ok, I will do that! 👍 

In the meantime, I noticed a minor issue (it's really about finishing touches here): if I'm a new player and hover the ?, I'm getting the help which is great but maybe I don't have enough time to read it all before it fades, so I would like to make it appear again. My instinct is to click on the ? (as my mouse is already hovering it), but this doesn't work. I have to move my mouse away and return to the ?. Could you simply react to the click too?

**Comment by armin-reichert** [comment]
Done. The original cause for using an image view with mouse event handlers instead of a Button was that a graphic button is broken in GWT, I should have "opened a ticket" for this.

**Comment by salmonb** [comment]
That was quick! Yes it's working now 👍  I will have a look later at the graphic button.

I will continue after lunch for the subdomains... 🍽️ 

Was wondering for new players: is there a way to know the cheat keys? 😄

**Comment by armin-reichert** [comment]
That's the essence of cheat keys! You get this information only on dubious sites in the darknet ;-)

**Comment by salmonb** [comment]
Back from lunch...

Good idea the fav icons. You can put them in the resources/public folder (either in `pacman-ui-fx` or `pacman-ui-fx-gwt`module).

For the code changes regarding the domains, should I create a separate branch or push them directly in the webfx branch?

**Comment by salmonb** [comment]
I pushed the changes in the `webfx-domains` branch. You can merge it to `webfx` whenever you want.

**Comment by armin-reichert** [comment]
Thanks. (I was a couple of hours out of house, short trip to Luxembourg).

I merged your change and  added the fav icons but my first try adding them int the GTW module's public resource folder did not work. I added the icons now to both folders, let's see if this works).

The two subdomains are working fine.

**Comment by salmonb** [comment]
Great for the subdomains! 😃 

I think the fav icons must be named favicon.ico by default. Otherwise, the favicon path must be set in index.html.

But I realise now that you want a different icon for each domain... So this solution is too simple. Maybe the best solution would be that WebFX supports Stage.getIcons() and map that to html automatically. I will think about it...

Otherwise I noticed that the cheat keys don't work anymore! Except immunity which is activated by Alt-Shift-N instead of Alt-Shit-I on my keyboard! Is it a bug introduced by our last keyCode trick?

**Comment by armin-reichert** [comment]
Yes, you are right about the favicons. I think we could just use the same favicon for both domains, so renaming one of them to "facicon.ico" should be sufficient. 

The cheat shortcuts work fine here (Windows 10, Goggle Chrome and Brave browser).  Alt-Shit-I toggles immunity, Alt-Shift-A toggles autopilot etc. 

May be a Mac keyboard (does it have an "Alt" key or is it this Mac "bathtub" key?) generates different key codes in JavaFX?

**Comment by salmonb** [comment]
Ok, lets keep it simple for now with the favicons 👍 

I think I found the issue with the cheat shortcuts. It's a NPE with letterCode, because `KeyCode.getKeyCode(letter)` may return null, especially when combined with modifier keys (maybe this doesn't happen on Windows, but it does on Mac). So can you add `letterCode != null && !letterCode.equals(currentEvent.getCode())` in your condition?

**Comment by armin-reichert** [comment]
Fix submitted.

**Comment by salmonb** [comment]
Fix working 👍 

For the favicon, I was wrong when assuming the browser will try favicon.ico by default (it was the case in the past, but apparently the browsers don't do that anymore). So it must be explicitly present in the html file. Since index.html is generated by WebFX, you need to add this section in your webfx.xml (either in `pacman-ui-fx` or `pacman-ui-fx-gwt` or module):

```xml

    <html>
        <head>
            <link rel="icon" type="image/ico" href="pacman.ico"/>
        </head>
    </html>

```

Change the href value with the icon you want, and then run `webfx update`.

**Comment by armin-reichert** [comment]
Thanks. However, already after changing the icon name to "favicon.ico", the icon (black ghost) appeared in the browser tab of each game.

**Comment by salmonb** [comment]
Right, I tried with Safari & Firefox and it works. Not in Chrome (my default browser), but maybe it's just a caching issue (will check again tomorrow).

**Comment by salmonb** [comment]
hi Armin,

I didn't have the time to make a publish announcement, but anyway Friday is not the best day for this, so I will do it beginning of next week if you're happy with that.

Strange but my Chrome was still not loading the favicon today. After different tries (you will see my commits) changing the format to png finally solved the issue (the ico format was working on my local machine but not on the live demo for any reason...). Then maybe you can work on that png file to fix the transparency issue you were talking about?

A last thing (because I thought otherwise you will start getting bored 😉) I was wondering if you could make it work on devices with no keyboard. I can't play it on my iPad for example. If you make it playable with a mouse, it will work also on mobiles because JavaFX automatically transforms not consumed touch events into mouse events. So basically it's about starting the game with just a mouse click, and then dragging pacman. You can see how they did that (the drag) in SpaceFX ([here](https://github.com/webfx-demos/webfx-demo-spacefx/blob/778a57ffcaccaad3e922db69cf54643aa975d538/webfx-demo-spacefx-application/src/main/java/eu/hansolo/spacefx/SpaceFXView.java#LL248C9-L248C9) exactly). The principle is to put JavaFX node (the touch area) over the character you want to control in the canvas and then react to the drag events. What do you think?

**Comment by armin-reichert** [comment]
Hello Bruno,

yes, I am completely fine with that. I just pushed anotherfavicon (no license restrictions, see https://thenounproject.com/icon/ghost-3109865/) and is also transparent (local test worked ok).

The non-keyboard steering is certainly a useful feature, but I have to first think about it: Pac-Man steering is (in my implementation) not just pressing the right key at the exact point where Pac-Man can change his direction. You can already preselect a direction before Pac-Man reaches the intersection and as soon as he can make the turn it will be done. 

If you do that with a drag-gesture, you will at least need some visual indication of the preselected move direction. If I press a key I have already a clear feedback of my input but when just wiping, how can I be sure what has been recognized? But I will have a look at the SpaceFX game about how this could be done.

**Comment by salmonb** [comment]
Maybe you can start with a simple `canvas.setOnMouseDragged(e -> {...})` and see what you can do? Maybe you can just calculate the delta between the mouse and pacman, and transform that into an equivalent arrow KeyEvent and pass it to your existing code? Just some ideas... Anyway, good luck with that 🍀

**Comment by armin-reichert** [comment]
What would normally be the replacement for a simple key press on a keyboard-less device like an iPad, buttons. (context) menus, gestures...?

I am sorry to ask such stupid questions but I have no idea how interaction with an iPad works in practice, even less inside a game.

**Comment by salmonb** [comment]
Reply: touch events, which are generated when touching the device with fingers, or with a pencil. But like I said, no need to use the JavaFX touch API, just make it work with the mouse, and it will work on touch devices as well. For example [SpaceFX](https://spacefx.webfx.dev) is playable with the mouse only, and therefore playable on touch devices.

**Comment by armin-reichert** [comment]
I added the following functionality (it was almost trivial): 

- Clicking somewhere on the intro screen now "inserts a coin" and changes to the credit screen.
- Clicking somewhere on the credit screen starts the game. 
- Making the ghosthouse to a "touchpad": Dragging the mouse / wiping from somewhere inside the ghosthouse to any direction steers Pac-Man to that direction.

I tested it with the mouse in the browser and in the desktop window and it looks ok to me. Could you please check if it works on your iPad too?

**Comment by salmonb** [comment]
I'm glad that it was not that difficult for you, and yes, I just played my first game on the iPad with a pencil!!! 👏 

Here is my feedback of this first play on iPad: 

I didn't make a big score as it was not the most intuitive experience for me. I instinctively tried to drag pacman and that didn't respond as I would expect. I was touching the screen with my pencil without releasing it, and that didn't work. Then I understood it was reacting to swipe events and not to drag events (more "work" for the player to make these gestures compared to just dragging). So I could play again, but then I was confused when it stopped working in the upper part. Then I read your instructions and understood... But I think most people will be a bit confused with that as well.

What's the idea of limiting the touch area to the lower part only? Is it to simulate the joystick in the original game machine? In this case maybe displaying a virtual joystick somewhere would clarify that, but it's much more work for you than just removing this limit and covering the whole canvas.

Can you also try to react to drag events, like in SpaceFX (you can see that it's not necessary to release the mouse to drag the spaceship)? I think this would lead to a much more intuitive experience. What do you think?

**Comment by armin-reichert** [comment]
Good morning Bruno,
 
as I had no touchpad to test I used the mouse to simulate a touchpad "wipe" gesture. 
 
Using the mouse, to get a defined motion like "Move UP", I need
 
- the start of gesture (mouse press)
- course of gesture (mouse drag)
- end of gesture (mouse release)
 
So I listen to a mouse press (gesture start), check if any drag event follows (not just a mouse click), and then listen to mouse release (end of gesture). Then I check the position change the mouse has made between gesture start and gesture end, and based on that decide which move direction this represents. That worked very well with the mouse, so I thought (as you mentioned) this would also translate well into touch events.
 
First, I only listened to drag events, and with a single gesture I got a list of these but I did not see how this list of drag events could be used to identify the gesture start/end and the user's intention.
 
 
My thought to restrict the touch pad to the lower part of the maze was that the user would not cover his sight with his hand, so wiping on the bottom should help with that.
 
But unfortunately this seems not to work so well with a real touchscreen :-(
 
Regards
Armin
 
 


  _____  

Von: Bruno Salmon ***@***.*** 
Gesendet: Sonntag, 21. Mai 2023 10:34
An: webfx-project/webfx-cli
Cc: Armin Reichert; Author
Betreff: Re: [webfx-project/webfx-cli] I don't quite get how this works (Issue #8)



I'm glad that it was not that difficult for you, and yes, I just played my first game on the iPad with a pencil!!! 👏

Here is my feedback of this first play on iPad:

I didn't make a big score as it was not the most intuitive experience for me. I instinctively tried to drag pacman and that didn't respond as I would expect. I was touching the screen with my pencil without releasing it, and that didn't work. Then I understood it was reacting to swipe events and not to drag events (more "work" for the player to make these gestures compared to just dragging). So I could play again, but then I was confused when it stopped working in the upper part. Then I read your instructions and understood... But I think most people will be a bit confused with that as well.

What's the idea of limiting the touch area to the lower part only? Is it to simulate the joystick in the original game machine? In this case maybe displaying a virtual joystick somewhere would clarify that, but it's much more work for you than just removing this limit and covering the whole canvas.

Can you also try to react to drag events, like in SpaceFX (you can see that it's not necessary to release the mouse to drag the spaceship)? I think this would lead to a much more intuitive experience. What do you think?

**Comment by salmonb** [comment]
I see your concern with the hand, but it's actually ok, the finger doesn't cover much the rest of the screen, and the palm is far enough from pacman. And with a pencil, it's even better.

You can actually detect the gestures in the 4 directions without the need to release the mouse/finger/pencil, just by comparing the relative position of the mouse with pacman. If I want pacman to go up, I just move my pencil above pacman. Same with other directions. Exactly like SpaceFX.

**Comment by armin-reichert** [comment]
Maybe I misunderstand something, but I am not convinced that Pac-Man should be steered by using the Pac-Man sprite or any point on the screen as the interaction area. I would rather use some dedicated are on the screen to similate a touchpad or mabe even better, the joystick from the original game. It would net be necessary to make it look like a joystick but rather to get the 5 joysticks states (neutral, up, down, left, right).

**Comment by salmonb** [comment]
Yes it's 2 different options. But like I said, the joystick option is more work for you - and a bit less intuitive in my opinion. You don't like the [SpaceFX](https://spacefx.webfx.dev/) game experience (just keep your mouse pressed to steer the spaceship)?

**Comment by armin-reichert** [comment]
The steering of the startship is certainly adequate in that game, but I have my doubts if this makes sense for Pac-Man. Pac-Man moves automatically without user interaction and only if there is a choice for changing direction (in my version in contrast to the original, already before reaching that point), only then user-interaction is necessary. That's a real difference to the slowly moving, almost static starship.

**Comment by salmonb** [comment]
I agree the games are different as pcaman can't just follow the mouse like the starship, due to the walls, and he can also continue walking if I release the mouse as opposed to SpaceFX.

That said I still think you can have an intuitive drag-based steering interaction even with these additonnal constraints. In the same way that I can press the up arrow key at any time to communicate my intention to move pacman up - even if not yet possible at the moment - I can communicate this same intention with the mouse by dragging it above pacman. It will move pacman up only once possible of course. I can eventually change my mind and communicate another direction in the meantime. In summary I don't see why it's not possible to communicate our intention with the mouse like we do with the keyboard. You could even internally transform these drag events into keyboard events for your game engine.

It's just my opinion, and if you prefer the joystick option, that's fine of course.

**Comment by armin-reichert** [comment]
Could you please create a new branch where I can try out different approaches? 

The problem I see with your proposal is that to e.g. move UP, you have to drag the mouse above Pac-Man's current position (correct?), so the interaction area becomes the complete scene and you always have to move around with your finger or pencil. I cannot judge how this feels in reality but I think it might not the best solution.

**Comment by salmonb** [comment]
What you describe as non-intuitive (following pacman with your mouse/finger/pencil) is exactly what I describe as the most intuitive!  Different opinions... 😆 

Ok, I created a new branch called `webfx-drag-steering` 👍

**Comment by armin-reichert** [comment]
I pushed a change (directly into the webfx branch) that now uses mouse drag+mouse release to steer Pac-Man. The gesture is recognized all over the game scene, not just in the lower part. Could you please check if this works on an iPad? With the mouse, it works well, in the browser and in a desktop window.

I also found an interesting blog post (from 2009): 
https://www.cxpartners.co.uk/our-thinking/touch_screen_gaming

**Comment by salmonb** [comment]
Yes, I confirm it's much better over the whole screen 👍 

But you still need to constantly release the mouse (or remove your finger or pencil) each time you want to change directions. It's quite annoying. A much better user experience would happen if you keep tracking the direction in the `setOnMouseDragged()` method, so we can steer Pac-Man for a long time all along a single press. The direction needs to be determined by comparing the current mouse position with the Pac-Man position (and not the initial dragging point). `setOnMouseDragged()` is called many times (basically each time you move the mouse, finger or pencil), but can call your direction consumer only when you detect a direction change.

I would be happy to help with the coding, but I have no idea how to get the Pac-Man position...

**Comment by armin-reichert** [comment]
Hi Bruno,

thanks for testing.

I don't understand why it is annoying to release the mouse  (or lift the finger off the touchscreen) at the end of a gesture. Isn't this totally normal for a "swipe" gesture? Why is it better to have the finger on the screen permanently? When using the mouse, it is certainly more comfortable to release the mouse button when a gesture ends than permanently pressing the button.

But I am only an occasional user of tablets and have *never* played any game on a tablet. So I trust your experience. Feel free to add your code , I am perfectly ok with that.

You can get the Pac-Man position as follows (when inside a GameScene):

```
context.game().level().ifPresent(level -> {
  var pos = level.pac().position();
``´`

`context` is a data structure that stores scene-related data: a reference to the game controller and its state, the current game model, the current level object and the references to Pac-Man and the ghosts etc.

The Pac-Man position is the left-upper corner of Pac-Man's "collision box" which is one square tile large (8x8 pixel). The sprite however has a size of 16 pixels. The position is in model-coordinates (unscaled).

**Comment by salmonb** [comment]
Thanks for sharing the code that gets the Pac-Man position 👍 

I made the change and pushed it to the webfx-drag-steering branch. To me, it's a much more enjoyable game experience, on the iPad for sure, and I think also on the desktop with the mouse. I let you try and please let me know what you think.

If you're happy with it, I let you finish the code (comment, cosmetics, etc...) as I did it quickly, and push in the webfx branch.

**Comment by salmonb** [comment]
hi Armin,

Just to let you know, I made a final improvement and pushed it to that webfx-drag-steering branch. The improvement is to react also to the intermediate directions (ex: you can lead Pac-Man to right-up direction with the mouse). I'm quite satisfied now with this drag steering. What do you think about it?

**Comment by armin-reichert** [comment]
I merged your changes, looks fine to me. Another thing: I wanted to push a workaround for the missing TinyLog logging library, but obviosuly I don't have write permission to webfx-demo-pacman-core. The change looks like this (Logger.java):
```
package org.tinylog;

public class Logger {

    private static void println(String message, Object... args) {
        message = message.replaceAll("\\{[^}]*\\}", "__"); // two underscores
        for (int i = 0; i < args.length; ++i) {
            message = message.replaceFirst("__", String.valueOf(args[i]));
        }
        var dateFormat = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss SSS");
        var prefix = LocalDateTime.now().format(dateFormat) + ": ";
        System.out.println(prefix + message);
    }

    public static void trace(String message, Object... parameters) {
        println(message, parameters); }

    public static void info(String message, Object... parameters) {
        println(message, parameters);
    }

    public static void error(String message, Object... parameters) {
        println(message, parameters);
    }
}

```

As the Logger is used by some tests that are run during the build, I can see that the code does something useful. Could you add it if it's ok for you?

**Comment by salmonb** [comment]
Great, it works very well on my iPad, thank you!

I gave you write access to webfx-demo-pacman-core, so feel free to do any change you like.

FYI `System.out.println()` doesn't print in the browser console. If you want logs in the browser as well, you can use `dev.webfx.platform.console.Console.log()` instead.

**Comment by salmonb** [comment]
hi Armin, do you think the game is now ready to go?

**Comment by armin-reichert** [comment]
Hi Bruno,

I would propose that you do the announcement on whatever channel you prefer. 

(I hate BigTech companies that censor free speech like Meta, Google, Twitter, etc.)

**Comment by salmonb** [comment]
Ok, I will do a tweet then (not a big fan of Twitter, but that's the channel most developers use to follow the project).

In the tweet, I will mention the 2 games urls and also give the link to the [webfx-demo-pacman](https://github.com/webfx-demos/webfx-demo-pacman/tree/webfx) repo. Can you upload a screenshot in the "Social Preview" section of the repo [settings](https://github.com/webfx-demos/webfx-demo-pacman/settings)?  It's to have a nice image at the bottom of the tweet, otherwise it will be an ugly generic GitHub repo image.

Also, are you ok if I mention your name as the author of these games? I'm asking in case you would prefer not to.

**Comment by armin-reichert** [comment]
Yes go ahead. I will add the image in GitHub.

**Comment by armin-reichert** [comment]
The (cropped) image should contain Ms. Pac-Man. Like thie attached one. Can you try if this looks better?

![social-preview](https://github.com/webfx-project/webfx-cli/assets/7394275/11a314a1-1a3e-44f5-878e-4683677507d7)

**Comment by armin-reichert** [comment]
I would also prefer the text "Ms. Pac-Man" instead of "Ms. version".

**Comment by salmonb** [comment]
Ok I updated my Tweet draft to say "Ms. Pac-Man" instead of "Ms. version" (this just reached the max number of characters for a tweet!), and uploaded your image in the GitHub repo (looks better I agree), also updated the repo description to say "This the WebFX version of the JavaFX games Pac-Man / Ms. Pac-Man written by Armin Reichert (see webfx branch)".  But that doesn't change the Twitter image yet (I guess they cached the image).

I hope the Twitter image will be refreshed tomorrow. End of day is not the best time for a professional tweet anyway. So I think I will wait tomorrow lunch time to send it.

> BTW: What does the sentence "First JavaFX... made by a WebFX user" mean? Are the other games made by WebFX non-users?

Difficult to be clear as we can't provide much details in a tweet... What I would like to highlight is that you closely participated to the WebFX port as opposed to other games authors, where we just did the whole port ourself without involving them (so we can't say they are WebFX users like you). But any clearer (yet short) wording is welcome.

**Comment by armin-reichert** [comment]
There is a German proverb "Gut Ding will Weile haben" ("good things need their time") so yes, don't hurry with that Tweet. 

Honestly, I don't feel like I helped so much with the downport. If I had known that the code must be Java 11 compatible, I could have saved you lots of (unnecessary) effort. In the meantime I have donwported the `pacman-core` repository in my main branch back to Java 11 (to ease future webfx ports which I then want to be able to do 100% by myself :-)

So let's finish the last details tomorrow. Thank you very much for all the work you have put into this.

**Comment by salmonb** [comment]
The idea behind that little sentence is also to encourage people to start experimenting WebFX, by saying: hey look what Armin did with WebFX! Yes of course I provided support, but still you are the first real WebFX user. Because as you know WebFX is still in incubation, and people are just watching the project for now. But hopefully at some point, they will git it a try too.

Otherwise yes, good advise, let's finish tomorrow 👍

**Comment by salmonb** [comment]
Twitter was still keeping the old image from its cache this morning... But by just adding / at the end of the repo url, this made Twitter react and update the image! 🥳  I also changed the wording a little bit:

<img width="572" alt="Screenshot 2023-05-24 at 11 11 04" src="https://github.com/webfx-project/webfx-cli/assets/15129227/a99b6a8d-1d83-4830-b6ef-2d7b1635c394">

What do you think?

**Comment by salmonb** [comment]
No space for additions... so maybe we can make a separate tweet to @elon 😉 

Great, so looks like we are good to go! Will just wait lunch time, so will send it in 1h.

Just a detail: the clickable links in your webfx README are the old ones (not working anymore). But not sure people will notice that.

**Comment by armin-reichert** [comment]
Links and README text updated.

**Comment by salmonb** [comment]
Tweet live: https://twitter.com/WebFXProject/status/1661336335427182593 🥳

**Comment by armin-reichert** [comment]
So it's now the Namco lawyers turn 😁

**Comment by salmonb** [comment]
As an immediate response of this tweet, Dirk Lemmermann [tweeted](https://twitter.com/dlemmermann/status/1661354440291500033) that he created a new [WebFX section](https://www.jfx-central.com/tools/webfx) in JFX-Central! With links to all WebFX demos - so including your games 😃 

It's also pretty sure that they will talk about your games in their [Links of the Week](https://www.jfx-central.com/home) on Friday 😉

**Comment by armin-reichert** [comment]
That's the international breakthrough for WebFX. Netflix series will be next! 😁

**Comment by salmonb** [comment]
Yes! So be prepared to be offered a role!! 😄 

Really good idea. Let me know when you will have a JavaFX version ready. For sure we can work on that 👍 

And what about integrating this mazes generator back to your Pac-Man game then?
And make a new version of the game with more mazes?
Jr. Pac-Man!! 🕹️

**Comment by armin-reichert** [comment]
Hi Bruno,

did you get my mail concerning my new WebFX version of Pac-Man?

**Comment by salmonb** [comment]
hi Armin!

No, I didn't receive the notification of your email for any reason!...

Glad that you made a new version 😃

I will have a look at the issues, and keep you updated 👍

**Comment by armin-reichert** [comment]
Great! The issue with the images is solved (I forgot to set the autoload parameter).
